pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -80,14 +58,14 @@ func (n Node) Compatible(pod *v1.Pod) error {
 		return err
 	}
 
+	tightened := n.Constraints.Requirements.Add(podRequirements.Requirements...)
```

Comment:
It feels like we're tripping over ourselves a bit. This is a tight loop that scales linearly with pods. We're already adding these requirements together inside `n.Constraints.Requirements.Compatible(podRequirements)`, and then we're doing it again here. Effectively, we have two areas of the code base responsible for the same thing (e.g. checking if a pod can schedule to a node), but done at different levels of abstraction.

As you've rightly done, we don't want to push this logic into the API package to create a dependency between `v1alpha5 -> cloudprovider`. Instead, if we could move the compatibility logic up to this layer, I *think* a lot of this impedance mismatch goes away. Obviously, we're blocked on this right now due to https://github.com/aws/karpenter/pull/1549, but it might be worth exploring once that's merged. 

I'm happy to merge this PR as it is currently, but would love to remove this rough edge for simplicity/performance.

Reply:
I was debating if the initial compatibility check was required when I wrote this and ended up with the conclusion that I wasn't certain that our Add() was written to handle possibly incompatible requirements, so it felt safer to keep it in.

Reply:
Just verified that the outer compatibility check handles the stuff that's not instance related (e.g. hostname requirements to implement topology spread).

Reply:
Not currently. You'd need to smoosh em together :D